[Elijah Ng] iP

0xEljh opened the pull request:
Add a task list that is maintained by the user and can contain todos, events, and deadlines. These contain descriptors of a task and relevant details (event details/deadline timings).
Tasks can be marked as completed by the user.
The added tasks can be displayed to the user, with indications of the type of task, its details, and whether the task has been completed.

0xEljh committed:
Add greet and exit

0xEljh committed:
Add Echo

User input is parsed and printed

0xEljh committed:
Add add, list

User input was only echoed back.

Update parses user input, maintains a list of tasks that can be retrieved on command.

0xEljh committed:
add done

User can use the done keyword to change the status of a task to done. List also keeps tracks of done and undone tasks.

0xEljh committed:
add Todo, Events, Deadlines

Main interfaced directly with TaskList and only Tasks could be created.

Introduced todos, events, and deadlines as new types of tasks that can be added with more details. Task manager now parses user input and puts a barrier between Duke and TaskList/other task implementations.

kaijiel24 commented on src/main/java/Task.java:
```diff
@@ -0,0 +1,28 @@
+    protected String description;
+    private boolean isComplete;
+
+    public Task(String description) {
+        this.description = description;
+        this.isComplete = false;
+    }
+
+    @Override
+    public String toString() {
+        //Subject to change; return string rep for printing purposes
+        return this.getCompletionStatusIcon() + " " + this.description; 
+    }
+
+    protected boolean markAsComplete() {
+        this.isComplete = true;
+        return this.isComplete;
+    }
+
+    protected boolean checkIsComplete() {
+        return this.isComplete;
+    }
+
+    private String getCompletionStatusIcon() {
+        return (isComplete ? "[\u2713]" : "[\u2718]");
+    }
```
Should you avoid using ```this``` if it is unnecessary, aka when it is clear that the variable is referring to the class element? It will also be good to keep it consistent as you used ```this.isComplete``` in ```checkIsComplete()``` but only ```isComplete``` in ```getCompletionStatusIcon```.  Other than this, well done!

kaijiel24 reviewed (commented):
Well done 👍

madbeez commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,127 @@
+        String keyword = inputs[0];
+
+        switch (keyword) {
+            case "bye":
+                return false;
+            case "list":
+                listTaskList();
+                return true;
+            case "done":
+                String msg = tasks.markComplete(Integer.parseInt(inputs[1]));
+                Duke.printBreak();
+                System.out.println("Nice! I've marked this task as done:");
+                System.out.println("     " + msg);
+                return true;
```
Perhaps you could store the boolean in a variable and return it at the end of the method, instead of returning in the middle of the switch statement.

madbeez commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,127 @@
+        ArrayList<String> detailWords = new ArrayList<>();
+        ArrayList<String> descriptionWords = new ArrayList<>();
+        
+        for (String input : userInput) {
+            if (flag) {
+                detailWords.add(input);
+                continue;
+            }
+
+            /** 
+             * Dev note: the following commented out code assumes that / is the keyword
+             * and not /at
+            if (input.charAt(0) == '/' || flag) {
+                detailWords.add(input.substring(1));
+                flag = true;
+                continue;
+            }
+             */
+
+            if (input.equals("/at")){
+                flag = true;
+                continue;
+            }
+            descriptionWords.add(input);
+        }
```
This section would be easier to read if the commented out code was placed at the end instead of in the middle.

madbeez reviewed (commented):
I like how you organised your code, and it was generally quite readable.

0xEljh committed:
Error Handling on TaskManager

TaskManager now informs user of errors due to erroneous user input: Missing command words, not enough arguments, and invalid int/index.

0xEljh committed:
Keywords increased

"at" and "by" keywords added to compliment existing keywords which had "/" prefixes.

0xEljh committed:
Merge pull request #1 from mxksowie/branch-level-5

Branch level 5

0xEljh committed:
Add delete

Delete task from task list by providing its index.

0xEljh committed:
Add Output class

Output class removes cyclic dependency of TaskManager and Duke. Output class will be used to further streamline the TaskManager class so that it does not need to handle parsing, commands, and outputs all together.

manaswini05 commented on src/main/java/TaskList.java:
```diff
@@ -0,0 +1,70 @@
+    @Override
+    public String toString() {
+        String result = "";
+        //for (Task task : tasks) {
```
Try to delete out the commented code if you do not plan on using it

manaswini05 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,151 @@
+        return true;
+    }
+    
+    private void deleteTask(int i) {
```
are you yet to implement this?

manaswini05 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,151 @@
+
+    private void markComplete(int i) {
+        try {
+            String msg = tasks.markComplete(i);
```
try using full words for string names

manaswini05 commented on src/main/java/TaskManager.java:
```diff
@@ -0,0 +1,151 @@
+
+    }
+
+    private void markComplete(int i) {
```
it is better if alphabet 'i' is used only as a loop counter
consider using another name for the integer

manaswini05 reviewed (commented):
Great work, esp on the comments 👍

0xEljh committed:
add Save

List of tasks now saved automatically to user's home directory after the list is updated.
This list is retrieved and loaded upon (re)starting the application.

0xEljh committed:
Merge pull request #2 from mxksowie/branch-Level-6

Branch level 6

0xEljh committed:
Merge pull request #3 from mxksowie/branch-Level-7

add Save and Retrieve tasks

0xEljh committed:
cleanup Main

Refactored code in main to improve readability

0xEljh committed:
Update TaskManager to have no prints

Streamlined methods in TaskManager by having Output class handle all printing. Extracted repeated/similar functionality in addEvent and addDeadline

0xEljh committed:
add Find based on keyword match

User can input find <keyword(s)> to get a list of tasks that contain the keywords (case insensitive)

0xEljh committed:
Add JavaDocs

Provide documentation for most non-trivial methods and classes

0xEljh committed:
Merge pull request #4 from mxksowie/branch-Level-9

add Find based on keyword match

0xEljh committed:
Merge branch 'master' into branch-A-JavaDoc

0xEljh committed:
Merge pull request #5 from mxksowie/branch-A-JavaDoc

Add JavaDocs

0xEljh committed:
Set theme jekyll-theme-slate

0xEljh committed:
Update user guide

0xEljh committed:
Update user guide

make code blocks more consistent

0xEljh committed:
add DateTime support

Event and deadline now support ISO dates as details.
New class DateType is new parent class to deadline and event to support this

0xEljh committed:
add date matching

Find (dated) tasks with matching dates when user queries for a date in ISO format.

0xEljh committed:
Merge pull request #6 from mxksowie/branch-Level-8

Branch level 8

0xEljh committed:
update user guide with date functionality
